lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,650 @@
+        // and return the complete payload.
+
+        // We need 10 bytes to read the message id, sequence number and total chunks
+        if chunk.remaining() < 10 {
+            let src_display = format!("{chunk:?}");
+            warn!(message = "Received malformed chunk headers (message ID, sequence number and total chunks) with less than 10 bytes. Ignoring it.",
+                src = src_display,
+                remaining = chunk.remaining(),
+                internal_log_rate_limit = true
+            );
+            return Ok(None);
+        }
+        let message_id = chunk.get_u64();
+        let sequence_number = chunk.get_u8();
+        let total_chunks = chunk.get_u8();
+
+        if total_chunks == 0 || total_chunks > GELF_MAX_TOTAL_CHUNKS {
+            warn!(
+                message = "Received a chunk with an invalid total chunks value. Ignoring it.",
+                message_id = message_id,
+                sequence_number = sequence_number,
+                total_chunks = total_chunks,
+                internal_log_rate_limit = true
+            );
+            return Ok(None);
+        }
+
+        if sequence_number >= total_chunks {
+            warn!(
+                message = "Received a chunk with a sequence number greater than total chunks. Ignoring it.",
+                message_id = message_id,
+                sequence_number = sequence_number,
+                total_chunks = total_chunks,
+                internal_log_rate_limit = true
+            );
+            return Ok(None);
+        }
+
+        let mut state_lock = self.state.lock().unwrap();
+
+        if state_lock.len() >= self.pending_messages_limit {
+            warn!(
+                message = "Received a chunk but reached the pending messages limit. Ignoring it.",
+                message_id = message_id,
+                sequence_number = sequence_number,
+                pending_messages_limit = self.pending_messages_limit,
+                internal_log_rate_limit = true
+            );
+            return Ok(None);
+        }
+
+        let message_state = state_lock.entry(message_id).or_insert_with(|| {
+            // We need to spawn a task that will clear the message state after a certain time
+            // otherwise we will have a memory leak due to messages that never complete
+            let state = Arc::clone(&self.state);
+            let timeout = self.timeout.clone();
+            let timeout_handle = tokio::spawn(async move {
+                tokio::time::sleep(timeout).await;
+                let mut state_lock = state.lock().unwrap();
+                if let Some(_) = state_lock.remove(&message_id) {
+                    // TODO: record metrics here? it would be insteresting to know how many messages are being discarded
+                    let message = format!("Message was not fully received within the timeout window of {}ms. Discarding it.",
```

Comment:
same as before, it would be very nice to know how many messages are timed out...